arangod/Replication2/ReplicatedLog/LogCommon.h

```diff
@@ -0,0 +1,315 @@
+    return !(right <= left);
+  }
+  [[nodiscard]] friend auto operator>=(T const& left, S const& right) -> bool {
+    return right <= left;
+  }
+  [[nodiscard]] friend auto operator>(T const& left, S const& right) -> bool {
+    return right < left;
+  }
+};
+
+struct LogIndex : implement_compare<LogIndex> {
+  constexpr LogIndex() noexcept : value{0} {}
+  constexpr explicit LogIndex(std::uint64_t value) noexcept : value{value} {}
+  std::uint64_t value;
+
+  [[nodiscard]] auto saturatedDecrement(uint64_t delta = 1) const noexcept -> LogIndex;
+
+  [[nodiscard]] auto operator<=(LogIndex) const -> bool;
+
+  [[nodiscard]] auto operator+(std::uint64_t delta) const -> LogIndex;
+
+  friend auto operator<<(std::ostream& os, LogIndex const& idx) -> std::ostream&;
+
+  [[nodiscard]] explicit operator velocypack::Value() const noexcept;
+};
+
+auto operator<<(std::ostream& os, LogIndex const& idx) -> std::ostream&;
+
+struct LogTerm : implement_compare<LogTerm> {
+  constexpr LogTerm() noexcept : value{0} {}
+  constexpr explicit LogTerm(std::uint64_t value) noexcept : value{value} {}
+  std::uint64_t value;
+  [[nodiscard]] auto operator<=(LogTerm) const -> bool;
+
+  friend auto operator<<(std::ostream& os, LogTerm const& term) -> std::ostream&;
+
+  [[nodiscard]] explicit operator velocypack::Value() const noexcept;
+};
+
+auto operator<<(std::ostream& os, LogTerm const& term) -> std::ostream&;
+
+[[nodiscard]] auto to_string(LogTerm term) -> std::string;
+[[nodiscard]] auto to_string(LogIndex index) -> std::string;
+
+struct TermIndexPair : implement_compare<TermIndexPair> {
+  LogTerm term{};
+  LogIndex index{};
+
+  friend auto operator<=(TermIndexPair const& left, TermIndexPair const& right) noexcept
+      -> bool;
+
+  TermIndexPair(LogTerm term, LogIndex index) noexcept;
+  TermIndexPair() = default;
+
+  void toVelocyPack(velocypack::Builder& builder) const;
+  [[nodiscard]] static auto fromVelocyPack(velocypack::Slice) -> TermIndexPair;
+
+  friend auto operator<<(std::ostream& os, TermIndexPair const& pair) -> std::ostream&;
+};
+
+[[nodiscard]] auto operator<=(TermIndexPair const& left, TermIndexPair const& right) noexcept -> bool;
+auto operator<<(std::ostream& os, TermIndexPair const& pair) -> std::ostream&;
+
+struct LogPayload {
+  LogPayload() = delete;
+  explicit LogPayload(velocypack::UInt8Buffer dummy);
+
+  // Named constructors, have to make copies.
+  [[nodiscard]] static auto createFromSlice(velocypack::Slice slice) -> LogPayload;
+  [[nodiscard]] static auto createFromString(std::string_view string) -> LogPayload;
+
+  friend auto operator==(LogPayload const&, LogPayload const&) -> bool;
+  friend auto operator!=(LogPayload const&, LogPayload const&) -> bool;
+
+  [[nodiscard]] auto byteSize() const noexcept -> std::size_t;
+  [[nodiscard]] auto slice() const noexcept -> velocypack::Slice;
+
+  // just a placeholder for now
+  velocypack::UInt8Buffer dummy;
```

Comment:
Please remember that a `velocypack::Buffer` always needs at least 216 bytes in the object itself, since it has a 192 byte builtin buffer to avoid extra allocations. This might or might not be what we need here. Just mentioning this, because we have fallen prey to this in the AgencyStore, where it ultimately lead to a huge memory usage overhead for small data.

Reply:
We now have `SmallBuffer` in `arangod/Agency/Node.h` to alleviate this.

Reply:
Good point, agreed, and we do have that on our radar. However, we haven't definitively decided how the payload should look like, as it's not really used yet, and thus stuck with that for now.